### ISSUE TITLE
Filtering Gold Label in the SRL Model

### ISSUE THREAD

**Issue by jasonyanwenl** [asyml/forte#324]
**Describe the bug**
In this [part](https://github.com/asyml/forte/blob/master/forte/models/srl/model.py#L275) of code, it says that only when both `span_idx` and `predicate_idx` are not `None`, we will assign gold label id in the corresponding position in tensor `gold_labels`.  Those two conditions mean:
* `span_idx` not `None` means that for a given gold link, the gold predicate argument span also found in model prediction.
* `predicate_idx` not `None` means that for a given gold link, the gold predicate also found in model prediction.

Why should those conditions be met? What if a gold label does not meet either of those two conditions? The gold label still exists and I think we should still assign the gold label id in the correspondinng position in `gold_labels`?

Thanks!

**Comment by hunterhector** [comment]
@huzecong Any ideas?

**Comment by huzecong** [comment]
The code was written more than a year ago so my memory's a bit blurry. I think the code in question was implementing this paper: https://www.aclweb.org/anthology/P18-2058.pdf. In section 2, paragraph "Beam pruning", it only keeps a portion ($\lambda_a$ and $\lambda_p$) of the candidate arguments and predicates before computing the loss. This is implemented by this block of code:
https://github.com/asyml/forte/blob/66462e7ce2d3b7138449c0272d7a66f4e342cad6/forte/models/srl/model.py#L394-L410

`_filter_labels`'s job is to discard gold semantic roles whose argument span or predicate was dropped by beam pruning, since we no longer have the logits for those span or predicates.

Does this make sense?

**Comment by jasonyanwenl** [comment]
> The code was written more than a year ago so my memory's a bit blurry. I think the code in question was implementing this paper: https://www.aclweb.org/anthology/P18-2058.pdf. In section 2, paragraph "Beam pruning", it only keeps a portion ($\lambda_a$ and $\lambda_p$) of the candidate arguments and predicates before computing the loss. This is implemented by this block of code:
> https://github.com/asyml/forte/blob/66462e7ce2d3b7138449c0272d7a66f4e342cad6/forte/models/srl/model.py#L394-L410
> 
> `_filter_labels`'s job is to discard gold semantic roles whose argument span or predicate was dropped by beam pruning, since we no longer have the logits for those span or predicates.
> 
> Does this make sense?

Hi @huzecong, thank you for your replying! Yeah, that makes more sense to me. So what you mean is that since prediction is limitied within top K, we should also limit the gold labels to be within top K. 

However, I am still little bit wondering about this. Beam pruning is only applied to prediction so why we should limit gold labels as well? This will cause some gold labels outside top K lost and model cannot learn that part of gold labels? In my sense, this could be like:

* We do beam pruning for prediction but build a tensor for all candidate (including candidate not within top K). Put top K predictions into related positions in this tensor.
* We build another tensor representing all the gold labels (not only top K).
* The loss will calculated over those two tensors?

Thanks again!

**Comment by huzecong** [comment]
> However, I am still little bit wondering about this. Beam pruning is only applied to prediction so why we should limit gold labels as well?

Hmm, I remember having the same doubts as you when I was implementing this. It wasn't obvious from the paper whether pruning was only applied during inference, so I checked the author's implementation and found out that it's applied both during training and inference:
https://github.com/luheng/lsgn/blob/b1495ecf559e36c5e37078c54e00a1ec6faae5b2/srl_model.py#L99-L108
I'm not sure whether this makes sense though, and I'd encourage you to try not pruning during training and see how that affects performance.

Regarding your suggestion, I'm not sure I completely understand --- is your proposal just to not prune candidates during training? If so then I think it's reasonable.

**Comment by jasonyanwenl** [comment]
> > However, I am still little bit wondering about this. Beam pruning is only applied to prediction so why we should limit gold labels as well?
> 
> Hmm, I remember having the same doubts as you when I was implementing this. It wasn't obvious from the paper whether pruning was only applied during inference, so I checked the author's implementation and found out that it's applied both during training and inference:
> https://github.com/luheng/lsgn/blob/b1495ecf559e36c5e37078c54e00a1ec6faae5b2/srl_model.py#L99-L108
> I'm not sure whether this makes sense though, and I'd encourage you to try not pruning during training and see how that affects performance.
> 
> Regarding your suggestion, I'm not sure I completely understand --- is your proposal just to not prune candidates during training? If so then I think it's reasonable.

Thanks for getting back to me! I will check the link you shared.

**Comment by hunterhector** [comment]
I think this is a very interesting issue. At a glance, this relates to multiple problems:

1. Exposure bias: If the training step and test step are different, for example, in training you always have the gold candidates in TOP-K, then it means in training you will not experience the same set of errors test time. This is called the exposure bias.

2. If during training, some gold labels are never seen due to beam search, the model will receive less training.

The method to reduce 2 can be using so-called "Teacher-forcing", where we always include the gold label. However, using "Teacher-forcing" may also mean we may encounter more exposure bias.

To avoid the problem of "teacher forcing" we can also do variations of "curriculum learning" (say curriculum beam) [1].

If we don't use "Teacher-Focing", another way to do this may be the "delayed LaSO" [1]

Another technique in searching structure is to use "Search Optimization", or Learning with Search Optimization (LaSO) [2,3].   

Overall, I don't think you can solve the problem easily just by adding gold labels into the Beam.

[1] https://www.aclweb.org/anthology/P14-1005.pdf
[2] https://arxiv.org/pdf/1606.02960.pdf
[3] https://arxiv.org/pdf/1606.02960.pdf
